components/raftstore-v2/src/operation/command/admin/split.rs

```diff
@@ -64,20 +74,387 @@ pub struct SplitResult {
+        // The locks of new regions will be put into the corresponding new regions
+        // later. And the locks belonging to the old region will stay in the original
+        // map.
+        let region_locks = {
+            let mut pessimistic_locks = self.txn_ext().pessimistic_locks.write();
+            info!(self.logger, "moving {} locks to new regions", pessimistic_locks.len(); "region_id"=> region_id);
+            // Update the version so the concurrent reader will fail due to EpochNotMatch
+            // instead of PessimisticLockNotFound.
+            pessimistic_locks.version = derived_epoch.get_version();
+            pessimistic_locks.group_by_regions(&regions, derived)
+        };
+
+        // Roughly estimate the size and keys for new regions.
+        let new_region_count = regions.len() as u64;
+        let estimated_size = self.approximate_size().map(|v| v / new_region_count);
+        let estimated_keys = self.approximate_keys().map(|v| v / new_region_count);
+        let mut meta = store_ctx.store_meta.lock().unwrap();
+        meta.set_region(
+            derived.clone(),
+            self,
+            RegionChangeReason::Split,
+            tablet_index,
+        );
+        self.post_split();
+
+        // It's not correct anymore, so set it to false to schedule a split
+        // check task.
+        self.set_may_skip_split_check(false);
+
+        let is_leader = self.is_leader();
+        if is_leader {
+            self.set_approximate_size(estimated_size);
+            self.set_approximate_keys(estimated_keys);
+            // todo: this may should be doing in handle_peer_split_response.
+            self.heartbeat_pd(store_ctx);
+
+            info!(
+                self.logger,
+                "notify pd with split";
+                "region_id" => self.region_id(),
+                "peer_id" => self.peer_id(),
+                "split_count" => regions.len(),
+            );
+
+            // todo: report to PD
+        }
+
+        assert!(self.split_progress_mut().is_empty());
+        let last_key = enc_end_key(regions.last().unwrap());
+        if meta.region_ranges.remove(&last_key).is_none() {
+            panic!("{:?} original region should exist", self.logger.list());
+        }
+        let last_region_id = regions.last().unwrap().get_id();
+        for (new_region, locks) in regions.into_iter().zip(region_locks) {
+            let new_region_id = new_region.get_id();
+
+            if new_region_id == region_id {
+                let not_exist = meta
+                    .region_ranges
+                    .insert(enc_end_key(&new_region), new_region_id)
+                    .is_none();
+                assert!(not_exist, "[region {}] should not exist", new_region_id);
+                continue;
+            }
+
+            let mut raft_message = self.prepare_raft_message();
+            raft_message.mut_from_peer().set_id(raft::INVALID_ID);
```

Comment:
Question: why need to set INVALID_ID

Reply:
This message is used to make store create an uninitialized peer. We can use it to distinguish from the msg sent from the leader in other stores.
![image](https://user-images.githubusercontent.com/71589810/200722993-0b9c78b9-5143-4417-8ba0-052fa2615ace.png)
![image](https://user-images.githubusercontent.com/71589810/200723015-73397446-454a-4333-ad76-d9197c82e7ed.png)